src/bsp/src/stm32/src/Decawave.cpp

```diff
@@ -0,0 +1,307 @@
+#include "Decawave.h"
+#include <DecawaveUtils.h>
+#include <FreeRTOS.h>
+#include <Task.h>
+#include <cstring>
+#include <deca_device_api.h>
+#include <deca_regs.h>
+#include <task.h>
+
+void rxCallback(const dwt_cb_data_t* callbackData, void* context) {
+    memcpy(&(static_cast<Decawave*>(context)->m_callbackData), callbackData, sizeof(dwt_cb_data_t));
+
+    BaseType_t taskWoken = pdFALSE;
+
+    if (static_cast<Decawave*>(context)->m_rxTaskHandle != NULL) {
+        vTaskNotifyGiveFromISR(static_cast<Decawave*>(context)->m_rxTaskHandle, &taskWoken);
+    }
+
+    static_cast<Decawave*>(context)->m_rxTaskHandle = NULL;
+    portYIELD_FROM_ISR(taskWoken);
+}
+
+void isrCallback(void* context) {
+    deca_selectDevice(static_cast<Decawave*>(context)->m_spiDevice);
+    dwt_isr();
+}
+
+void rxAsyncTask(void* context) {
+    while (true) {
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+        static_cast<Decawave*>(context)->retrieveRxFrame(
+            static_cast<Decawave*>(context)->m_rxFrame);
+    }
+}
+
+Decawave::Decawave(decaDevice_t spiDevice) :
+    m_spiDevice(spiDevice),
+    m_channelNo(2),
+    m_speed(UWBSpeed::SPEED_110K),
+    m_rxAsyncTask("dw_rx_task", tskIDLE_PRIORITY + 10, rxAsyncTask, this) {}
+
+Decawave::Decawave(decaDevice_t spiDevice, uint8_t channel, UWBSpeed speed) :
+    m_spiDevice(spiDevice),
+    m_speed(speed),
+    m_rxAsyncTask("dw_rx_task", tskIDLE_PRIORITY + 10, rxAsyncTask, this) {
+
+    if (channel > 0 && channel < 8) {
+        m_channelNo = channel;
+    }
+}
+
+bool Decawave::start() {
+    deca_selectDevice(m_spiDevice);
+    deca_setSlowRate();
+    uint32_t deviceID = 0;
+    uint8_t i = 0;
+    // Retry to read the deviceID 10 times before abandoning
+    while (deviceID != DWT_DEVICE_ID && i++ < 10) {
```

Comment:
Do while would probably be legitimate here, also your i will be incremented once before doing the loop once, so it's actually 9 times I think

Reply:
Changed for do while